src/Components/Web.Extensions/src/InputFile/InputFile.cs

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Rendering;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Web.Extensions
+{
+    /// <summary>
+    /// A component that wraps the HTML file input element and exposes a <see cref="Stream"/> for each file's contents.
+    /// </summary>
+    public class InputFile : ComponentBase, IInputFileJsCallbacks, IDisposable
+    {
+        private ElementReference _inputFileElement;
+
+        private IJSRuntime _jsRuntime = default!;
+
+        private IJSUnmarshalledRuntime? _jsUnmarshalledRuntime;
+
+        private InputFileJsCallbacksRelay? _jsCallbacksRelay;
+
+        [Inject]
+        private IServiceProvider ServiceProvider { get; set; } = default!;
+
+        /// <summary>
+        /// Gets or sets the event callback that will be invoked when the collection of selected files changes.
+        /// </summary>
+        [Parameter]
+        public EventCallback<IFileListEntry[]> OnChange { get; set; }
+
+        /// <summary>
+        /// Gets or sets the maximum message size for file data sent over a circuit.
+        /// This only has an effect when using Blazor Server.
+        /// </summary>
+        [Parameter]
+        public int MaxMessageSize { get; set; } = 20 * 1024; // SignalR limit is 32K.
```

Comment:
Open question: to clarify the intent here, should we rename this to something like `MaxSignalRChunkSize`? The goal would be to clarify that:

 * ... this only applies in the SignalR case (Blazor Server) due to the nature of SignalR
 * ... that we're only limiting the size of chunks, not the size of the total upload

Please feel free to come up with a better name if you can!